### PR TITLE
ARROW-18186: [C++][MinGW] Make buildable with clang

### PR DESCRIPTION
**Pull request by kou**
Error1 (can't use `[[gnu::dllexport]]` with template):

    cpp/src/arrow/util/int_util.cc:463:1: error: an attribute list cannot appear here
    INSTANTIATE_ALL()
    ^~~~~~~~~~~~~~~~~
    cpp/src/arrow/util/int_util.cc:454:3: note: expanded from macro 'INSTANTIATE_ALL'
      INSTANTIATE_ALL_DEST(uint8_t)  \
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    cpp/src/arrow/util/int_util.cc:444:3: note: expanded from macro 'INSTANTIATE_ALL_DEST'
      INSTANTIATE(uint8_t, DEST)       \
      ^~~~~~~~~~~~~~~~~~~~~~~~~~
    cpp/src/arrow/util/int_util.cc:440:12: note: expanded from macro 'INSTANTIATE'
      template ARROW_TEMPLATE_EXPORT void TransposeInts( \
               ^~~~~~~~~~~~~~~~~~~~~
    cpp/src/arrow/util/visibility.h:47:31: note: expanded from macro 'ARROW_TEMPLATE_EXPORT'
    #define ARROW_TEMPLATE_EXPORT ARROW_DLLEXPORT
                                  ^~~~~~~~~~~~~~~
    cpp/src/arrow/util/visibility.h:32:25: note: expanded from macro 'ARROW_DLLEXPORT'
    #define ARROW_DLLEXPORT [[gnu::dllexport]]
                            ^~~~~~~~~~~~~~~~~~

Error2 (unused variable):

    cpp/src/arrow/util/io_util.cc:1079:7: warning: variable 'oflag' set but not used [-Wunused-but-set-variable]
      int oflag = _O_CREAT | _O_BINARY | _O_NOINHERIT;
          ^

Error3 (missing field initializers):

    cpp/src/arrow/util/io_util.cc:1545:29: warning: missing field 'InternalHigh' initializer [-Wmissing-field-initializers]
      OVERLAPPED overlapped = {0};
                                ^

### CODE REVIEW

File: .github/workflows/cpp.yml [hunk 1]
```diff
@@ -293,18 +293,21 @@ jobs:
...
-        mingw-n-bits:
-          - 32
-          - 64
+        include:
+          - msystem_lower: mingw32
+            msystem_upper: MINGW32
+          - msystem_lower: mingw64
+            msystem_upper: MINGW64
+          - msystem_lower: clang64
+            msystem_upper: CLANG64
```
Comment (pitrou) [inline .github/workflows/cpp.yml:310]: CI job took 85 minutes...

**Comment by pitrou** [comment]
PR looks fine a priori, but can we try to reduce those build times?

**Comment by kou** [comment]
> can we try to reduce those build times?

Should we do in this pull request? Or can we defer this to ARROW-18039 ?

**Comment by pitrou** [comment]
We can defer this to ARROW-18039, but I'd like to make sure this isn't forgotten :-)

**Comment by kou** [comment]
OK! I'll do it in this month!

**Pull request merged**